source/material_model/cpp.cc

```diff
@@ -0,0 +1,674 @@
+/*
+  Copyright (C) 2018 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/material_model/cpp.h>
+
+
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    template <int dim>
+    bool
+    CPP<dim>::
+    is_compressible () const
+    {
+      return compressible_param;
+    }
+
+    template <int dim>
+    double
+    CPP<dim>::
+    reference_viscosity () const
+    {
+      return reference_viscosity_param;
+    }
+
+    template <int dim>
+    void
+    CPP<dim>::
+    evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+             MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      if (needs_simulator)
```

Comment:
Yeah, I went a bit overboard (prematurely) optimizing things. The thought was to make every function pass the bare minimum number of parameters. I don't know if calling get_simulator has any cost associated with it, beyond what's already used just checking this conditional.

There are a bunch of places inside the generated cpp code that do something similar (e.g. generate different function declarations depending on what parameters are actually used). In that code it's probably even less necessary, since the compiler can see whether or not each parameter is actually used.

I wrote it both ways (see [commit 15df124](https://github.com/geodynamics/aspect/tree/15df124ca5e5424568d0eae5d47df4851ec1069d) ), and ended up going with the convoluted one because I spent time on it, and it works. But it definitely adds more complexity.

I'm open to either way, depending on what you and the other developers prefer.

Reply:
I also do that sometimes ;)

The function get_simulator, just returns the pointer to the simulator (see https://github.com/geodynamics/aspect/blob/master/source/simulator/simulator_access.cc#L58) without any checks, so my guess would be that that would not be very slow. I would except it to be something which could be very easily optimized by a compiler. 

On the other hand, an if/else statement is something which may slow down you code as well. That is because the cpu will probably have to wait till the comparison is done, or somehow speculate which branch is more likely to be executed, which might go wrong. 

But as you may see from the language I am using, it is in general very hard to know what will be faster without testing it. My intuition would actually say that using the if/else statement is slower, but I could be completely wrong ;)

I think that for the clarity of the code it would be good simplify it, and only if you know for sure that the more complex code is significantly faster, and it located in a place where it matters (which is the case here), use more complex code.